incubator/virtualcluster/experiment/pkg/scheduler/cache/cache.go

```diff
@@ -34,19 +37,64 @@ type schedulerCache struct {
 
 	mu sync.RWMutex
 
+	tenants    map[string]struct{}
 	clusters   map[string]*Cluster
 	pods       map[string]*Pod
 	namespaces map[string]*Namespace
 }
 
 func NewSchedulerCache(stop <-chan struct{}) Cache {
-	return &schedulerCache{
+	c := &schedulerCache{
 		stop:       stop,
+		tenants:    make(map[string]struct{}),
 		clusters:   make(map[string]*Cluster),
 		pods:       make(map[string]*Pod),
 		namespaces: make(map[string]*Namespace),
 	}
+	go wait.Until(c.GarbageCollection, 3*time.Minute, stop)
+	return c
+}
+
+func (c *schedulerCache) GarbageCollection() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	csToDelete := make([]*Cluster, 0)
+	for _, v := range c.clusters {
+		if v.shadow && metav1.Now().After(v.lastUpdateTime.Add(5*time.Minute)) {
+			csToDelete = append(csToDelete, v)
+		}
+	}
+	for _, each := range csToDelete {
+		delete(c.clusters, each.name)
+	}
+}
+
+func (c *schedulerCache) AddTenant(n string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.tenants[n] = struct{}{}
+}
+
+func (c *schedulerCache) RemoveTenant(n string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 
+	nsToDelete := make([]*Namespace, 0)
```

Comment:
simply  `var nsToDelete []*Namespace` is better. this is a nil pointer while make create a zero value